### PR TITLE
feat: Native Dockerfile

### LINKED ISSUE #579: Switch to Ansible-free Docker image
In order to advance implementation of both [OEP-45: Configuring and Operating Open edX](https://open-edx-proposals.readthedocs.io/en/latest/architectural-decisions/oep-0045-arch-ops-and-config.html) and https://github.com/openedx/public-engineering/issues/51, we would like to switch devstack from using primarily Docker images built with Ansible code from the configuration repository to images built from Dockerfiles in each service's repository. We hope this will also help simplify Tutor by providing better base images to derive from, with fewer workarounds and duplicated code blocks required in Tutor.

Please create sub-tasks in separate issues as appropriate, in this repository and/or each service needing changes to its Dockerfile or image upload process. Services should be switched one at a time so the work can be done and tested incrementally. The process should look something like the following for each service:

Ensure that the repo's Dockerfile can create a base image which is appropriate for small production installations
Ensure that the repo's Dockerfile also defines an additional image derived from the base which adds dependencies and configuration changes needed for development environments
Check with the Tutor developers during code review to see if any further changes to the Dockerfile would help simplify Tutor
Automate uploads of both images when code changes are merged
Switch devstack to use the new Ansible-free development image, and do some basic testing to ensure it works before merging

### PR DESCRIPTION
**Pull request by aht007**
ISSUE: https://github.com/openedx/edx-analytics-data-api/issues/579
This PR is part of effort aimed at removing Ansible based configurations and replacing them with `Dockerfile`. Currently Devstack Docker images are built using `Ansible based configurations` in the `configurations` repository. Through this effort we will make sure that the Repo has its own `Dockerfile` which has all the necessary configurations to setup small production and dev environments.

**Steps to run this Image with Devstack:**

- Build the Image locally first using the target `dev` i.e. `docker build -t analytics_api_dev --target dev .`
- Checkout to the branch in this PR of `devstack` https://github.com/openedx/devstack/pull/986
- Run `make dev.provision.analytics_api` to run the provisioning script for the new image.
- Run `make dev.up.analytics_api` in the terminal while in the devstack directory.

### CODE REVIEW

**Comment by aht007** [comment]
@openedx/masters-devs-cosmonauts I can see that the current Dockerfile of this repository is being used for CI purposes and is also being pushed to `openedx` docker hub account. I would need your team's guidance as to what extent that image is useful and do we need to maintain that image as well or we can just retire that and use our new docker image for CI testing as well.
Besides that I would also like for your team to take a look at the changes in this PR.

File: requirements/dev.in
Comment (schenedx) [inline requirements/dev.in:4]: Can you explain why this is needed now with a native docker file? Is it the case that Ansible was taking care of this dependency?

File: Dockerfile
Comment (schenedx) [inline Dockerfile:77]: We are no longer hooking up with newrelic-admin with this change. Why would you want to leave that out?

File: .github/workflows/push-docker-images.yml
Comment (schenedx) [inline .github/workflows/push-docker-images.yml:41]: Why is this commented out? Is it the case we don't really push this image to any repositories?

**Review by schenedx — COMMENTED** [review]
I got inline comments.

File: Dockerfile
Comment (alangsto) [inline Dockerfile:6]: it looks like not all of these packages were copied over in the new format, is that intentional (i.e. `libffi-dev`)

File: requirements/dev.in
Comment (aht007) [reply requirements/dev.in:4]: Yeah, in ansible the production file was also being installed along with the dev requirements file. That's not what we want now as dev target should only install dev requirements file, but when I tried to do this the container threw errors while connecting to mysql because `mysqlclient` was missing. That's why I added it to `dev.in` instead of installing the `production.txt` in dev image.

File: Dockerfile
Comment (aht007) [reply Dockerfile:77]: We don't use new relic in `devstack`. As for the current case I haven't added it to prod target as well since according to our discussion with Jeremy we don't want to make `new relic` a necessary part of our base images by default. We can make a separate target for that in our Dockerfile or leave it to our orchestration system(k8s or docker-compose) to do those things for us.

File: .github/workflows/push-docker-images.yml
Comment (aht007) [reply .github/workflows/push-docker-images.yml:41]: I have commented that out on purpose. Since our initial target is `devstack` so I am just pushing the dev images for now. Also the old Ansible based `devstack` image is pushed to `edxops/analytics_api` repository and enabling this push will override that image too. For smooth transition, I am pushing the Native Images to new repository(`analytics_api_dev`) and keeping the old image intact so if we want to switch over to older image due to some issues we always can.

File: Dockerfile
Comment (aht007) [reply Dockerfile:6]: Yeah thats intentional. I couldn't find any use of these packages e.g `gettext`, `libffi-dev`, `git-core`, `openjdk` and others so I removed them.

File: requirements/dev.in
Comment (schenedx) [reply requirements/dev.in:4]: Thanks for that explanation. This make sense.

File: Dockerfile
Comment (schenedx) [reply Dockerfile:77]: Got it. Do you have wiki page documents talk about this? I'd like this point to be written down as a policy for layering newrelic to our apps.

File: .github/workflows/push-docker-images.yml
Comment (schenedx) [reply .github/workflows/push-docker-images.yml:41]: If you are doing this on purpose, then the whole section should be wrapped with comments to help others understand why this is left in the code, but commented out. Please provide the explanation in the code to help others understand the desire for the smooth transition.

File: Dockerfile
Comment (schenedx) [reply Dockerfile:6]: @aht007 Could you explain to us how you arrived at the conclusion that these are not used? Did you bring up the service and test through the newly built docker image?

Comment (aht007) [reply Dockerfile:6]: It is both testing the new image as well as learning from other images as well. These packages were also present in Insights' Ansible roles but I removed them from there as well since I couldn't find their usage. Also as per our knowledge many of these packages are present in Ansible roles and docker files for historical reasons but they aren't used now so we are better off without them.
I have tried bringing up the container and testing the image and it looked fine but if someone from your team knows about any of the removed package's usage then I will be happy to put it back there.

**Review by schenedx — APPROVED** [review]
👍

**Comment by aht007** [comment]
@schenedx Regarding the old docker publish workflow, I searched for that image's usage and I couldn't find it anywhere, not in this repo and not in the `openedx` org at all. That workflow was added as a part of CI overhaul when we shifted to Github Actions and I think that image was being pushed for just historical reasons without any usage. I have removed that workflow and related commands. Also the CI was using old Ansible based image which I have replaced with the newer image and CI is running fine with that(I tested the push in this PR so the new image is available on edxops). 
I would again need one final review from your side on this PR as well this [PR of devstack](https://github.com/openedx/devstack/pull/986) which introduces the newer image usage.

**Comment by schenedx** [comment]
@aht007 Reviewed your devstack PR as well as re-review on this one. Still LGTM.
Feel free to merge.

**Pull request merged**
